### ISSUE TITLE
surprising result when updating variable in WHILE with EXPR

### ISSUE THREAD

**Issue by Luis-Henriquez-Perez** [okamsn/loopy#18]
When creating a while loop with loopy where the EXPR contains the same variable as that of the WHILE--in this case `rest`, the loop never runs because rest is set to nil at the beginning of the loop. I wrote I want a loop that continues based on some dynamic circumstance.

 This surprised me but I get why it happens. Is there some way for loopy to account for this case and not bind REST to nil initially?

If there is not a good way of intelligently accounting for this, I want to add this to the wiki. Maybe in a "gotcha"/"potentially suprising behaviors" section.

```elisp
(defun xl-define-key::clause (rest)
  "Convert clause into proper general-def arguments."
  (loopy ((while rest)
	  (list identify-fn xl-define-key::identifiers-alist)
	  (expr (clause rest) (funcall identify-fn rest)))))
```

```elisp
(let
      ((clause nil)
       (rest nil) ; => set to nil
       (identify-fn nil))
    (let*
	((list-395 xl-define-key::identifiers-alist))
      (cl-block nil
	(cl-tagbody
	 (while
	     (consp list-395)
	   (if rest nil
	     (go loopy--non-returning-exit-tag))
	   (setq identify-fn
		 (car list-395))
	   (setq rest
		 (funcall identify-fn rest))
	   (setq clause
		 (pop rest))
	   (setq rest
		 (car rest))
	   (setq list-395
		 (cdr list-395)))
	 loopy--non-returning-exit-tag)
	nil)))
```

I realize this could be fixed by doing the following but it seems too wordy.

```elisp
(defun xl-define-key::clause (rest)
  "Convert clause into proper general-def arguments."
  (loopy ((while rest)
	  (list identify-fn xl-define-key::identifiers-alist)
	  (expr (clause new-rest) (funcall identify-fn rest))
	  (do (setq rest new-rest)))))
```

**Comment by okamsn** [comment]
One way to handle this is to use the `with` macro argument, effectively doing `(let* ((rest  rest)) ...`. Currently, the package checks whether a variable is already defined by `with` or another loop command before binding.

``` elisp
(defun test (test-var)
       (loopy (with (test-var test-var))
              ((expr test-var (cdr test-var))
               (return test-var))))

(cl-assert (equal (test '(1 2 3)) '(2 3)))
```
Another way, probably clearer, is to add a macro argument that specifies variables to not initialize. Maybe something like "`without`" or "`no-init`".

``` elisp
(let ((a 1) (b 2))
  (loopy (without a b)
         (with (c 3))
         ((expr a (+ a c))
          (expr b (+ b c))
          (return a b)))) ; => '(4 5)
```
Do you think that "`without`" is an understandable word for this meaning? I think that it contrasts well with "`with`", but that it also relies on the meaning of "`with`".

**Comment by Luis-Henriquez-Perez** [comment]
> Do you think that "without" is an understandable word for this meaning?

Yes, I think the name `without` is a great name.

**Comment by okamsn** [comment]
`without` has been added as a keyword, with the alias `no-init`. Does this sufficiently solve the problem?

**Comment by Luis-Henriquez-Perez** [comment]
Yes. I'll close this. Thank you.
